### ISSUE TITLE
[RFC] Add Top-nσ logit truncation example via custom logits processor

### ISSUE THREAD

**Issue by Jackie2049** [vllm-project/vllm#45023]
## RFC: Add Top-nσ Logit Truncation Example via Custom Logits Processor

**Reference**: Tang et al., "Top-nσ: Not All Logits Are You Need", ACL 2025. [arXiv:2411.07641](https://arxiv.org/abs/2411.07641)

### Motivation

Top-nσ is a logit-space dynamic truncation method proposed by Tang et al. (ACL 2025). Unlike probability-space filters such as `top_p` or `min_p`, it applies **before softmax** and keeps the candidate set less sensitive to temperature changes.

This issue proposes adding a **research-oriented example implementation** using vLLM's existing custom logits processor API. It is **not** a request to add a new built-in sampling parameter.

### Scope

This proposal only adds an **example file** demonstrating how to implement top-nσ as a custom logits processor. It:
- Does **not** add a new `SamplingParams` field
- Does **not** change any existing sampling behavior
- Does **not** modify the core `Sampler` class

### Algorithm

For each request row:

```
1. max_logit = logits.max()
2. std_logit = logits.std()
3. threshold = max_logit - n * std_logit
4. logits[logits < threshold] = -inf   (mask unlikely tokens)
5. Let existing vLLM sampler handle temperature, softmax, and sampling
```

The standard deviation directly measures the "peakiness" of the logit distribution:
- **Sharp distribution** (model is certain): small std → narrow threshold → fewer candidates
- **Flat distribution** (model is uncertain): large std → wide threshold → more candidates

### How it compares to other truncation methods

| Method | Space | Truncation rule | Adaptive? |
|---|---|---|---|
| top_k | Probability | Keep top-k highest-probability tokens | ❌ Fixed count |
| top_p (nucleus) | Probability | Keep tokens until cumulative probability ≥ p | ✅ Adaptive count only |
| min_p | Probability | Keep tokens with prob ≥ max_prob × min_p | ✅ Relative threshold |
| top_nσ | **Logit** | threshold = max_logit − n × std | ✅ Dual-adaptive |
| temperature | Logit | logits / T — rescales distribution steepness | ❌ No truncation |

> Key distinction: top_k/top_p/min_p operate **after softmax** (probability space); top_nσ operates **before softmax** (logit space).

### Why Top-nσ?

- **Truncation before softmax** — top_nσ filters in logit space first, then softmax only processes the surviving candidates (~2.5% of vocab). In contrast, top_p/min_p must compute a full-vocab softmax first, then truncate in probability space.
- **Dual adaptivity** — the std-based threshold adjusts both *how many* and *how loosely* candidates are kept in one formula: sharp distributions auto-narrow, flat ones auto-widen. top_k fixes the count; top_p adapts the count but not the threshold tightness.
- **Temperature-stable** — candidate count stays consistent across temperature scaling (logit std scales proportionally), while top_p/min_p candidate counts vary significantly with T.
- **Argmax-invariant** — the top-1 token always survives (0 std from itself), so greedy decoding can skip this processor entirely.

### Usage (via custom logits processor)

```python
from vllm import SamplingParams

sampling_params = SamplingParams(
    temperature=0.8,
    extra_args={"top_n_sigma": 2.0}
)
```

The custom logit processor reads `top_n_sigma` from `extra_args` and applies truncation before the standard sampling pipeline.

### Safety / Edge Cases

| Case | Behavior |
|------|----------|
| `n <= 0` | Reject (raise `ValueError`) |
| `std == 0` (all logits equal) | Keep all tokens (no filtering) |
| NaN/Inf logits | Preserve existing behavior (no-op) |
| After filtering | **At least the argmax token is always kept** (prevent all-`-inf`) |

### Comparison Example

```python
# Logit distribution: max=10.0, std=2.0

top_k=5:           Keeps exactly 5 tokens
top_p=0.9:         Keeps tokens summing to 90% cumulative probability
min_p=0.05:        Keeps tokens with prob >= 5% of max
top_n_sigma=2.0:   Keeps tokens with logit >= 10.0 - 2*2.0 = 6.0
```

### Deliverables (if maintainers are interested)

1. **Example file**: `examples/features/logits_processor/top_n_sigma.py`
2. **Unit tests**: Verify masking, edge cases, and compatibility with other samplers
3. **Documentation**: One-paragraph note in sampling docs

---

**References**: Tang et al., "Top-nσ: Not All Logits Are You Need", ACL 2025. [arXiv:2411.07641](https://arxiv.org/abs/2411.07641)
